examples/step-52/step-52.cc

```diff
@@ -616,34 +631,33 @@ namespace Step52
                                     max_delta,
                                     refine_tol,
                                     coarsen_tol);
-    output_results(time, 0, method);
-
-    // Now for the time loop. The last time step is chosen such that the final
-    // time is exactly reached.
-    unsigned int n_steps = 0;
-    while (time < final_time)
+    output_results(initial_time, 0, method);
+    DiscreteTime time(initial_time, final_time, time_step);
+    while (time.is_at_end() == false)
       {
-        if (time + time_step > final_time)
-          time_step = final_time - time;
-
-        time = embedded_explicit_runge_kutta.evolve_one_time_step(
-          [this](const double time, const Vector<double> &y) {
-            return this->evaluate_diffusion(time, y);
-          },
-          time,
-          time_step,
-          solution);
+        const double new_time =
+          embedded_explicit_runge_kutta.evolve_one_time_step(
+            [this](const double time, const Vector<double> &y) {
+              return this->evaluate_diffusion(time, y);
+            },
+            time.get_current_time(),
+            time.get_next_step_size(),
+            solution);
+        time.set_next_step_size(new_time - time.get_current_time());
```

Comment:
I am not sure I understand why you do this: isn't the RK always going to return ``get_current_time+get_next_step_size``? I think this line is not necessary.

Reply:
This line is absolutely necessary since the function `evolve_one_time_step()` may reduce the current dt (in addition to the next dt) internally.
https://github.com/dealii/dealii/blob/ff89e12c96f2186d84e14dc3b7d798e2461710ce/include/deal.II/base/time_stepping.templates.h#L809-L815

Reply:
Ah, yes. It is a bit unfortunate that the 3 loops looks different now or what do you think?

Reply:
I have already added these comments before the aforementioned code to explain it better.

```c++
   // embedded method. In summary, time step size is potentially modified in
   // three ways:
   //   - Reducing or increasing time step size within
   //     TimeStepping::EmbeddedExplicitRungeKutta::evolve_one_time_step().
   //   - Using the calculated <code>delta_t_guess</code>.
   //   - Automatically adjusting the step size of the last time step to ensure
   //     simulation ends precisely at <code>final_time</code>. This adjustment
   //     is handled inside the DiscreteTime instance.
```